components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/publish/SynapseObservabilityHandler.java

```diff
@@ -0,0 +1,300 @@
+            } else if (null != synCtx.getProperty(SynapseConstants.IS_INBOUND)) {
+                String inboundEndpointName = synCtx.getProperty(PrometheusMetricsConstants.INBOUND_ENDPOINT_NAME).
+                        toString();
+                incrementErrorCount(inboundEndpointName);
+                stopTimers(synCtx, PrometheusMetricsConstants.INBOUND_ENDPOINT);
+            } else {
+                String context = axis2MessageContext.getProperty(PrometheusMetricsConstants.TRANSPORT_IN_URL).
+                                                                                                 toString();
+                String apiInvocationUrl = axis2MessageContext.getProperty(PrometheusMetricsConstants.SERVICE_PREFIX).
+                                                             toString() + context.replaceFirst(DELIMITER, EMPTY);
+                String apiName = getApiName(context, synCtx);
+                String remoteAddress = (String) axis2MessageContext.getProperty
+                                              (org.apache.axis2.context.MessageContext.REMOTE_ADDR);
+
+                incrementErrorCount(apiName, apiInvocationUrl, remoteAddress);
+                stopTimers(synCtx, PrometheusMetricsConstants.API);
+            }
+            synCtx.setProperty(PrometheusMetricsConstants.IS_ERROR, true);
+        }
+        return true;
+    }
+
+    @Override
+    public boolean handleRequestInFlow(MessageContext synCtx) {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx)
+                .getAxis2MessageContext();
+        String remoteAddress = (String) axis2MessageContext.getProperty(
+                org.apache.axis2.context.MessageContext.REMOTE_ADDR);
+
+        if (null == synCtx.getProperty(PrometheusMetricsConstants.REST_FULL_REQUEST_PATH)) {
+            if (null != synCtx.getProperty(SynapseConstants.PROXY_SERVICE)) {
+                String proxyName = axis2MessageContext.getAxisService().getName();
+
+                incrementCount(proxyName, remoteAddress);
+                startTimers(synCtx, proxyName, PrometheusMetricsConstants.PROXY_SERVICE, null);
+            } else if (null != synCtx.getProperty(SynapseConstants.IS_INBOUND)) {
+                String inboundEndpointName = synCtx.getProperty(SynapseConstants.INBOUND_ENDPOINT_NAME).toString();
+
+                incrementCount(inboundEndpointName);
+                startTimers(synCtx, inboundEndpointName, PrometheusMetricsConstants.INBOUND_ENDPOINT,
+                        null);
+            } else {
+                String context = axis2MessageContext.getProperty(PrometheusMetricsConstants.TRANSPORT_IN_URL).
+                        toString();
+                String apiInvocationUrl = axis2MessageContext.getProperty(PrometheusMetricsConstants.SERVICE_PREFIX).
+                        toString() + context.replaceFirst(DELIMITER, EMPTY);
+                String apiName = getApiName(context, synCtx);
```

Comment:
Could you check in current SP analytics , how we are getting this api name .. This processing seems too big